In [1]:
import numpy as np
import pandas as pd

In [2]:
train = pd.read_csv('data/raw/train.csv', parse_dates=['Date'])
test = pd.read_csv('data/raw/test.csv', parse_dates=['Date'])
store = pd.read_csv('data/raw/store.csv')

g:\wks\github\metanov\rossman-store-sales\venv\lib\site-packages\IPython\core\interactiveshell.py:2728: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
train.describe()

Store     DayOfWeek         Sales     Customers          Open  \
count  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06   
mean   5.584297e+02  3.998341e+00  5.773819e+03  6.331459e+02  8.301067e-01   
std    3.219087e+02  1.997391e+00  3.849926e+03  4.644117e+02  3.755392e-01   
min    1.000000e+00  1.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    2.800000e+02  2.000000e+00  3.727000e+03  4.050000e+02  1.000000e+00   
50%    5.580000e+02  4.000000e+00  5.744000e+03  6.090000e+02  1.000000e+00   
75%    8.380000e+02  6.000000e+00  7.856000e+03  8.370000e+02  1.000000e+00   
max    1.115000e+03  7.000000e+00  4.155100e+04  7.388000e+03  1.000000e+00   

              Promo  SchoolHoliday  
count  1.017209e+06   1.017209e+06  
mean   3.815145e-01   1.786467e-01  
std    4.857586e-01   3.830564e-01  
min    0.000000e+00   0.000000e+00  
25%    0.000000e+00   0.000000e+00  
50%    0.000000e+00   0.000000e+00  
75%    1.000000e+00   0.000000e+00  
max    1.000000e+00   1.000000e+00

In [4]:
train.head()

Store  DayOfWeek       Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5 2015-07-31   5263        555     1      1            0   
1      2          5 2015-07-31   6064        625     1      1            0   
2      3          5 2015-07-31   8314        821     1      1            0   
3      4          5 2015-07-31  13995       1498     1      1            0   
4      5          5 2015-07-31   4822        559     1      1            0   

   SchoolHoliday  
0              1  
1              1  
2              1  
3              1  
4              1

In [5]:
print("Store unique values: ", len(train['Store'].unique()))
print("Open unique values: ", len(train['Open'].unique()))
print("Promo unique values: ", len(train['Promo'].unique()))
print("StateHoliday unique values: ", len(train['StateHoliday'].unique()), "-> (",train['StateHoliday'].unique(), ")")
print("SchoolHoliday unique values: ", len(train['SchoolHoliday'].unique()), "-> (",train['SchoolHoliday'].unique(), ")")

Store unique values:  1115
Open unique values:  2
Promo unique values:  2
StateHoliday unique values:  5 -> ( ['0' 'a' 'b' 'c' 0] )
SchoolHoliday unique values:  2 -> ( [1 0] )


In [6]:
# We need to correct some badly interpreted StateHoliday values 0 -> "0"
train.loc[train['StateHoliday'] == 0, 'StateHoliday'] = "0"

In [7]:
test.describe()

Id         Store     DayOfWeek          Open         Promo  \
count  41088.000000  41088.000000  41088.000000  41077.000000  41088.000000   
mean   20544.500000    555.899533      3.979167      0.854322      0.395833   
std    11861.228267    320.274496      2.015481      0.352787      0.489035   
min        1.000000      1.000000      1.000000      0.000000      0.000000   
25%    10272.750000    279.750000      2.000000      1.000000      0.000000   
50%    20544.500000    553.500000      4.000000      1.000000      0.000000   
75%    30816.250000    832.250000      6.000000      1.000000      1.000000   
max    41088.000000   1115.000000      7.000000      1.000000      1.000000   

       SchoolHoliday  
count   41088.000000  
mean        0.443487  
std         0.496802  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         1.000000

In [8]:
test.head()

Id  Store  DayOfWeek       Date  Open  Promo StateHoliday  SchoolHoliday
0   1      1          4 2015-09-17   1.0      1            0              0
1   2      3          4 2015-09-17   1.0      1            0              0
2   3      7          4 2015-09-17   1.0      1            0              0
3   4      8          4 2015-09-17   1.0      1            0              0
4   5      9          4 2015-09-17   1.0      1            0              0

In [9]:
print("Store unique values: ", len(test['Store'].unique()))
print("Open unique values: ", len(test['Open'].unique()))
print("Promo unique values: ", len(test['Promo'].unique()))
print("StateHoliday unique values: ", len(test['StateHoliday'].unique()), "-> (",test['StateHoliday'].unique(), ")")
print("SchoolHoliday unique values: ", len(test['SchoolHoliday'].unique()), "-> (",test['SchoolHoliday'].unique(), ")")

Store unique values:  856
Open unique values:  3
Promo unique values:  2
StateHoliday unique values:  2 -> ( ['0' 'a'] )
SchoolHoliday unique values:  2 -> ( [0 1] )


In [10]:
store.describe()

Store  CompetitionDistance  CompetitionOpenSinceMonth  \
count  1115.00000          1112.000000                 761.000000   
mean    558.00000          5404.901079                   7.224704   
std     322.01708          7663.174720                   3.212348   
min       1.00000            20.000000                   1.000000   
25%     279.50000           717.500000                   4.000000   
50%     558.00000          2325.000000                   8.000000   
75%     836.50000          6882.500000                  10.000000   
max    1115.00000         75860.000000                  12.000000   

       CompetitionOpenSinceYear       Promo2  Promo2SinceWeek  Promo2SinceYear  
count                761.000000  1115.000000       571.000000       571.000000  
mean                2008.668857     0.512108        23.595447      2011.763573  
std                    6.195983     0.500078        14.141984         1.674935  
min                 1900.000000     0.000000         1.000000      2009.000000  
25%                 2006.000000     0.000000        13.000000      2011.000000  
50%                 2010.000000     1.000000        22.000000      2012.000000  
75%                 2013.000000     1.000000        37.000000      2013.000000  
max                 2015.000000     1.000000        50.000000      2015.000000

In [11]:
store.head()

Store StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0      1         c          a               1270.0                        9.0   
1      2         a          a                570.0                       11.0   
2      3         a          a              14130.0                       12.0   
3      4         c          c                620.0                        9.0   
4      5         a          a              29910.0                        4.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              NaN              NaN   
1                    2007.0       1             13.0           2010.0   
2                    2006.0       1             14.0           2011.0   
3                    2009.0       0              NaN              NaN   
4                    2015.0       0              NaN              NaN   

     PromoInterval  
0              NaN  
1  Jan,Apr,Jul,Oct  
2  Jan,Apr,Jul,Oct  
3              NaN  
4              NaN

In [12]:
print("Store unique values: ", len(store['Store'].unique()))
print("StoreType unique values: ", len(store['StoreType'].unique()))
print("Assortment unique values: ", len(store['Assortment'].unique()))

Store unique values:  1115
StoreType unique values:  4
Assortment unique values:  3


In [13]:
# Prior to do the join, we quickly modify the store dataframe to set 'Store' as index
store.set_index('Store', inplace=True)

In [14]:
# Let's do the join in both train and test dataframe 
train = train.join(store, on='Store')
test = test.join(store, on='Store')

In [15]:
train.isna().sum()

Store                             0
DayOfWeek                         0
Date                              0
Sales                             0
Customers                         0
Open                              0
Promo                             0
StateHoliday                      0
SchoolHoliday                     0
StoreType                         0
Assortment                        0
CompetitionDistance            2642
CompetitionOpenSinceMonth    323348
CompetitionOpenSinceYear     323348
Promo2                            0
Promo2SinceWeek              508031
Promo2SinceYear              508031
PromoInterval                508031
dtype: int64

In [16]:
test.isna().sum()

Id                               0
Store                            0
DayOfWeek                        0
Date                             0
Open                            11
Promo                            0
StateHoliday                     0
SchoolHoliday                    0
StoreType                        0
Assortment                       0
CompetitionDistance             96
CompetitionOpenSinceMonth    15216
CompetitionOpenSinceYear     15216
Promo2                           0
Promo2SinceWeek              17232
Promo2SinceYear              17232
PromoInterval                17232
dtype: int64

In [17]:
test[np.isnan(test['Open'])]

Id  Store  DayOfWeek       Date  Open  Promo StateHoliday  \
479      480    622          4 2015-09-17   NaN      1            0   
1335    1336    622          3 2015-09-16   NaN      1            0   
2191    2192    622          2 2015-09-15   NaN      1            0   
3047    3048    622          1 2015-09-14   NaN      1            0   
4759    4760    622          6 2015-09-12   NaN      0            0   
5615    5616    622          5 2015-09-11   NaN      0            0   
6471    6472    622          4 2015-09-10   NaN      0            0   
7327    7328    622          3 2015-09-09   NaN      0            0   
8183    8184    622          2 2015-09-08   NaN      0            0   
9039    9040    622          1 2015-09-07   NaN      0            0   
10751  10752    622          6 2015-09-05   NaN      0            0   

       SchoolHoliday StoreType Assortment  CompetitionDistance  \
479                0         a          c                  NaN   
1335               0         a          c                  NaN   
2191               0         a          c                  NaN   
3047               0         a          c                  NaN   
4759               0         a          c                  NaN   
5615               0         a          c                  NaN   
6471               0         a          c                  NaN   
7327               0         a          c                  NaN   
8183               0         a          c                  NaN   
9039               0         a          c                  NaN   
10751              0         a          c                  NaN   

       CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
479                          NaN                       NaN       0   
1335                         NaN                       NaN       0   
2191                         NaN                       NaN       0   
3047                         NaN                       NaN       0   
4759                         NaN                       NaN       0   
5615                         NaN                       NaN       0   
6471                         NaN                       NaN       0   
7327                         NaN                       NaN       0   
8183                         NaN                       NaN       0   
9039                         NaN                       NaN       0   
10751                        NaN                       NaN       0   

       Promo2SinceWeek  Promo2SinceYear PromoInterval  
479                NaN              NaN           NaN  
1335               NaN              NaN           NaN  
2191               NaN              NaN           NaN  
3047               NaN              NaN           NaN  
4759               NaN              NaN           NaN  
5615               NaN              NaN           NaN  
6471               NaN              NaN           NaN  
7327               NaN              NaN           NaN  
8183               NaN              NaN           NaN  
9039               NaN              NaN           NaN  
10751              NaN              NaN           NaN

The given store Nr. 622 is the only one with a NaN, and when looking at the other store, they are all opened. So, let's make the assumption that the store is opened on that day.

In [18]:
test.loc[np.isnan(test['Open']), 'Open'] = 1
test['Open'] = test['Open'].astype(int)

I will do a little bit of feature engineering as follows:
- refactor the date to add a `Day`, `Month` and `Year` column and drop the `Date` column.
- add a new column `isCompetition` out of `CompetitionOpenSinceMonth` and `CompetitionOpenSinceYear` and drop the latest. Also, set 0 all `CompetitionDistance` that are NaN.
- change column `Promo2` based on all `Promo2xxx` related field and drop those. Since this kind of promotion is constantly running because it is renewed every 3 months, only a proper time series model would probably capture its contribution.
This way, most of the NaN will be removed, apart from one that will be discussed in a moment.

In [19]:
train['Day'] = train['Date'].dt.day
train['Week'] = train['Date'].dt.week
train['Month'] = train['Date'].dt.month
train['Year'] = train['Date'].dt.year

test['Day'] = test['Date'].dt.day
test['Week'] = test['Date'].dt.week
test['Month'] = test['Date'].dt.month
test['Year'] = test['Date'].dt.year

def isCompetition(row):
    if row['Year'] >= row['CompetitionOpenSinceYear'] and \
       row['Month'] >= row['CompetitionOpenSinceMonth']:
        return 1
    else:
        return 0

train['isCompetition'] = train.apply(lambda row: isCompetition(row), axis=1)
test['isCompetition'] = test.apply(lambda row: isCompetition(row), axis=1)
train.loc[np.isnan(train['CompetitionDistance']), 'CompetitionDistance'] = 0
test.loc[np.isnan(test['CompetitionDistance']), 'CompetitionDistance'] = 0

def isPromo2(row):
    if row['Promo2'] == 1 and \
       row['Year'] >= row['Promo2SinceYear'] and \
       row['Week'] >= row['Promo2SinceWeek']:
        return 1
    else:
        return 0
        
train['Promo2'] = train.apply(lambda row: isPromo2(row), axis=1)
test['Promo2'] = test.apply(lambda row: isPromo2(row), axis=1)

In [20]:
drop_columns = ['Date', 'Week', \
                'CompetitionOpenSinceMonth', 'CompetitionOpenSinceYear', \
                'Promo2SinceWeek', 'Promo2SinceYear', 'PromoInterval']
train.drop(drop_columns + ['Customers'], axis=1, inplace=True)
test.drop(drop_columns, axis=1, inplace=True)

Finally, since we are sure that a closed shop is going to do 0 sales, we can strip the corresponding rows and remove the `Open` column. This way, the model will be able to focus only on real sales, and we will be able to standardize the Sales input properly.

In [21]:
train.drop(train[train['Open'] == 0].index, inplace=True)
train.drop(['Open'], axis=1, inplace=True)
test.drop(test[test['Open'] == 0].index, inplace=True)
test.drop(['Open'], axis=1, inplace=True)
test.insert(loc=2, column='Sales', value=0)

In [22]:
train.to_csv('data/processed/train.csv', index=False)
test.to_csv('data/processed/test.csv', index=False)